File: src/Partido.java
import java.text.NumberFormat;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Locale;

public class Partido implements Comparable<Partido> {

  private String sigla;
  private String numero;

  private int votosNominais;
  private int votosLegenda;
  private int quantidadeVotos;

  private LinkedList<Candidato> candidatos = new LinkedList<>();

  public void setCandidatos(LinkedList<Candidato> candidatos) {
    this.candidatos = candidatos;
  }

  public LinkedList<Candidato> getCandidatos() {
    return new LinkedList<>(this.candidatos);
  }

  public void insereCandidato(Candidato c) {
    this.candidatos.add(c);
  }

  public Partido(String numero, String sigla) {
    this.numero = numero;
    this.sigla = sigla;
  }

  @Override
  public int compareTo(Partido outroPartido) {
    int comparaVotos = Integer.compare(outroPartido.quantidadeVotos, this.quantidadeVotos);
    if (comparaVotos != 0)
      return comparaVotos;
    return this.numero.compareTo(outroPartido.numero);
  }

  public int getVotosNominais() {
    return votosNominais;
  }

  public int getVotosLegenda() {
    return votosLegenda;
  }

  public int getQuantidadeVotos() {
    return this.quantidadeVotos;
  }

  public int getQuantidadeCandidatosEleitos() {
    int count = 0;
    for (Candidato c : this.candidatos) {
      if (c.getEleito())
        count++;
    }
    return count;
  }

  public static Comparator<Partido> comparaPorCandidatoMaisVotado() {
    return new Comparator<Partido>() {
      @Override
      public int compare(Partido p1, Partido p2) {

        // Encontra a qtd de votos do candidato 1 mais votado
        int votosCandidato1 = 0;
        for (Candidato c : p1.getCandidatos()) {
          if (c.getQuantidadeVotos() > votosCandidato1) {
            votosCandidato1 = c.getQuantidadeVotos();
          }
        }
        // Encontra a qtd de votos do candidato 2 mais votado
        int votosCandidato2 = 0;
        for (Candidato c : p2.getCandidatos()) {
          if (c.getQuantidadeVotos() > votosCandidato2) {
            votosCandidato2 = c.getQuantidadeVotos();
          }
        }
        int comparaVotos = Integer.compare(votosCandidato2, votosCandidato1);
        if (comparaVotos != 0) {
          return comparaVotos;
        }
        return p1.getNumero().compareTo(p2.getNumero());
      }
    };
  }

  public void incrementaVotosNominais(int votos) {
    this.votosNominais += votos;
    this.quantidadeVotos += votos;
  }

  public void incrementaVotosLegenda(int votos) {
    this.votosLegenda += votos;
    this.quantidadeVotos += votos;
  }

  public String getSigla() {
    return this.sigla;
  }

  public void setSigla(String sigla) {
    this.sigla = sigla;
  }

  public String getNumero() {
    return this.numero;
  }

  public void setNumero(String numero) {
    this.numero = numero;
  }

  @Override
  public String toString() {
    NumberFormat brFormat = NumberFormat.getInstance(Locale.forLanguageTag("pt-BR"));
    String output = "";
    output += this.sigla + " - " + this.numero + ", " + brFormat.format(quantidadeVotos) + " votos ";
    output += "(" + brFormat.format(votosNominais) + " nominais" + " e " + brFormat.format(votosLegenda)
        + " de legenda), ";

    output += this.getQuantidadeCandidatosEleitos();
    if (this.getQuantidadeCandidatosEleitos() <= 1)
      output += " candidato eleito";
    else
      output += " candidatos eleitos";

    return output;
  }

  public String getRelatorioMaisEMenosCandidatoVotado() {
    Collections.sort(this.candidatos);
    Candidato maisVotado = this.candidatos.getFirst();
    Candidato menosVotado = this.candidatos.getLast();
    NumberFormat brFormat = NumberFormat.getInstance(Locale.forLanguageTag("pt-BR"));
    String output = this.sigla + " - " + this.numero + ", ";
    output += maisVotado.getNomeUrna() + " (" + maisVotado.getNumeroCandidato() + ", "
        + brFormat.format(maisVotado.getQuantidadeVotos()) + " votos) / ";
    output += menosVotado.getNomeUrna() + " (" + menosVotado.getNumeroCandidato() + ", "
        + brFormat.format(menosVotado.getQuantidadeVotos()) + " votos)";
    return output;
  }

}

File: implementacao-c++/partido.cpp
#include "partido.hpp"

void Partido::setSigla(const string& sigla) {
  this->sigla = sigla;
}

string Partido::getSigla() const {
  return this->sigla;
}

void Partido::setNumero(const string& numero) {
  this->numero = numero;
}

string Partido::getNumero() const {
  return this->numero;
}

void Partido::setVotosNominais(const int& votos) {
  this->votosNominais = votos;
}

int Partido::getVotosNominais() const {
  return this->votosNominais;
}

void Partido::setVotosLegenda(const int& votos) {
  this->votosLegenda = votos;
}

int Partido::getVotosLegenda() const {
  return this->votosLegenda;
}

void Partido::setQtdVotos(const int& votos) {
  this->qtdVotos = votos;
}

int Partido::getQtdVotos() const {
  return this->qtdVotos;
}

void Partido::setQtdCandidatosEleitos(const int& eleitos) {
  this->qtdCandidatosEleitos = eleitos;
}

int Partido::getQtdCandidatosEleitos() const {
  return this->qtdCandidatosEleitos;
}

void Partido::insereCandidato(const Candidato& candidato) {
  this->candidatos.push_back(const_cast<Candidato*>(&candidato));
}

const vector<Candidato*>& Partido::getCandidatos() const {
  return this->candidatos;
}

int Partido::comparaPartidos(const Partido& p1, const Partido& p2) const {
  if (p1.getQtdVotos() > p2.getQtdVotos())
    return 1;
  if (p1.getQtdVotos() < p2.getQtdVotos())
    return -1;
  return 0;
}

int Partido::comparaPartidosPorCandidatosMaisVotados(const Partido& p1, const Partido& p2) const {
  int votosC1 = 0;
  int votosC2 = 0;

  // Encontra a quantidade de votos do candidato mais votado do partido atual
  for (const auto& candidato : this->getCandidatos()) {
    if (candidato->getQuantidadeVotos() > votosC1) {
      votosC1 = candidato->getQuantidadeVotos();
    }
  }

  // Encontra a qtd de votos do candidato mais votado do partido 2
  for (auto candidato : p2.getCandidatos()) {
    if (candidato->getQuantidadeVotos() > votosC2)
      votosC2 = candidato->getQuantidadeVotos();
  }

  if (votosC1 > votosC2)
    return 1;
  if (votosC1 < votosC2)
    return -1;
  return 0;
}

void Partido::incrementaVotosNominais(const int& votos) {
  this->votosNominais += votos;
}

void Partido::incrementaVotosLegenda(const int& votos) {
  this->votosLegenda += votos;
}
